1178 Fix tag deletion / merge can mangle tags_cache

Taeir opened the pull request:
In my development database, it actually has a newline after every line, including the last one, so it seems this fix will work very well with hardly any changes.

Fixes #1178

trichoplax reviewed (changes requested):
A big improvement over the existing behaviour (I confirmed in my local development environment that under the existing code, deleting a prefix of a tag can result in the tags_cache no longer being interpreted as an array, so the site breaks when trying to edit a post that used to have the affected tag).

A question before approving:
1. I've added a comment about the existing overriding of the built in `exec` command. Should we address that here or raise a separate issue, or is it not a problem?

Points that don't need to be resolved here:
1. Merging a tag into another tag results in a post that previously had both tags now having one tag twice (it only shows once until editing, where it shows twice). This appears to be an existing problem so I see no reason to delay this pull request. I have raised a separate issue #1186 .
2. Both merging and deletion lead to hard deletion of the tag to be removed. This makes it also disappear from the edit history of every post it has ever been applied to.
3. This also means that the removed tag can be recreated by adding it to a post. Soft deletion would allow storing an indicator that the tag is not to be recreated, or which tag should be used instead (like a deleted synonym).

2 and 3 are existing problems that do not need to be addressed here. I have mentioned them in the discussion on #1177 in case they can be included in changes there or possibly a new issue.

Taeir commented on app/controllers/tags_controller.rb:
```diff
@@ -133,7 +133,7 @@ def merge
                 'SET posts.tags_cache = REPLACE(posts.tags_cache, ?, ?) ' \
                 'WHERE posts_tags.tag_id = ?'
-    exec([posts_sql, "\n- #{@subordinate.name}", "\n- #{@primary.name}", @subordinate.id])
+    exec([posts_sql, "\n- #{@subordinate.name}\n", "\n- #{@primary.name}\n", @subordinate.id])
```
Since exec is quite a dangerous command to run, I would be in favor of renaming the function to something that is at least not a ruby builtin accessible at all locations. Something like `query` or `exec_query` would probably be a better name.

ArtOfCode- replied:
Sounds sensible to me. @Taeir do you want to do that in this PR or merge this and do that separately?

Taeir replied:
I've updated the PR with the new name for `exec` (`exec_sql`)

Taeir commented:
@trichoplax
1. Addressed

Your other problems:
1. Addressed in #1193 
3. Tag merging should be editing the post history such that it looks like the new tag was added/removed, as it replaces references to the subordinate tag with the primary tag. As far as I can see, that should work fine. For tag deletion, you are right.
4. Perhaps we should be utilizing synonyms for this, such that the recreation scenario you describe becomes not possible.

trichoplax reviewed (changes requested):
@Taeir 

> 1. Addressed

I couldn't see a change from `exec` in this pull request, but I notice you have made the change in #1193 so now it won't be needed here.

We still use `exec` overriding the built in command in an unrelated file `scripts/wipe_everything.rb`. How would you like this to be addressed?
- In this pull request.
- In #1193 along with the existing change.
- As a separate issue (I'm happy to raise an issue if you prefer this option).

> Your other problems:
> 1. Addressed in #1193 
> 2. Tag merging should be editing the post history such that it looks like the new tag was added/removed, as it replaces references to the subordinate tag with the primary tag. As far as I can see, that should work fine. For tag deletion, you are right.
> 3. Perhaps we should be utilizing synonyms for this, such that the recreation scenario you describe becomes not possible.

1. I'll respond to this on #1193.
2. & 3. Synonyms sounds like a good solution, but I'll leave further discussion to #1177 and any new issues resulting from that, to avoid expanding scope here.

Taeir commented:
@trichoplax I have 4 different origins set up in git, and it pushed to the wrong one. should be updated here now.

For the wipe_everthing script, I made a separate PR (#1194).
